File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        n = len(prices)
        buy_prices = [float('inf')]*k
        max_profit = [0]*k
        
        for i in range(len(prices)):
            for j in range(k):
                buy_prices[j] = min(buy_prices[j], prices[i] if j == 0 else prices[i]-max_profit[j-1])
                max_profit[j] = max(max_profit[j], prices[i]-buy_prices[j])

        return max_profit[k-1]

File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        int[] buyPrice = new int[k];
        for (int i = 0; i < k; i++) {
            buyPrice[i] = Integer.MAX_VALUE;
        }
        int[] maxProfit = new int[k];
        for(int i = 0; i < prices.length; i++){
            for(int j = 0; j<k; j++) {
                if(j == 0) buyPrice[j] = Math.min(buyPrice[j], prices[i]);
                else buyPrice[j] = Math.min(buyPrice[j], prices[i]-maxProfit[j-1]);
                maxProfit[j] = Math.max(maxProfit[j], prices[i]-buyPrice[j]);
            }
        }
        return maxProfit[k-1];

        
    }
}